src/inmanta/parser/plyInmantaParser.py

```diff
@@ -736,49 +736,74 @@ def p_constant_mls(p: YaccProduction) -> None:
     attach_from_string(p)
 
 
-format_regex = r"""({{\s*([\.A-Za-z0-9_-]+)\s*}})"""
+format_regex = r"""({{(\s*([\.A-Za-z0-9_-]+)\s*)}})"""
 format_regex_compiled = re.compile(format_regex, re.MULTILINE | re.DOTALL)
 
 
-def get_string_ast_node(string: LocatableString, mls: bool) -> Union[Literal, StringFormat]:
-    matches = [[str(string)[m.start() : m.end()], m.start(), m.end()] for m in format_regex_compiled.finditer(str(string))]
+def get_string_ast_node(string_ast: LocatableString, mls: bool) -> Union[Literal, StringFormat]:
+    matches: List[re.Match[str]] = list(format_regex_compiled.finditer(str(string_ast)))
     if len(matches) == 0:
-        return Literal(str(string))
-
-    offset = 3 if mls else 1  # len(""")  or len(') or len(")
-    init: List[str]
-    match: List[str]
-    locatable_matches: List(LocatableString) = []
-    for obj in matches:
-        init, match, _ = (str(string)[start:end].splitlines() for start, end in pairwise((0, obj[1], obj[2], len(obj))))
-        start_line: int = string.lnr + len(init) - 1
-        end_line: int = start_line + len(match) - 1
-        len_last_init_line: int = len(init[-1]) if init else 0
-        start_char: int = len(init[-1]) + offset + string.start if len(init) == 1 else len_last_init_line + 1
-        end_char: int = len(match[-1]) + start_char - 1 if len(match) == 1 else len(match[-1])
-        range: Range = Range(string.location.file, start_line, start_char, end_line, end_char)
-        locatable_string = LocatableString("".join(match), range, string.lexpos, string.namespace)
-        locatable_matches.append(locatable_string)
-    return create_string_format(string, locatable_matches)
-
-
-def create_string_format(format_string: LocatableString, variables: List[LocatableString]) -> StringFormat:
+        return Literal(str(string_ast))
+
+    mls_offset: int = 3 if mls else 1  # len(""")  or len(') or len(")
+    locatable_matches: List[Tuple[str, LocatableString]] = []
+    for match in matches:
+        init_string: str = str(string_ast)[0 : match.start() + 2]
+        match_string: str = match[2]
+        init_lines: List[str] = (init_string + "\n").splitlines()  # lines before the match
+        match_lines: List[str] = match_string.splitlines()  # lines matching the string format
+        line_offset, char_offset = get_offset(match_lines, match[3])
```

Comment:
I don't understand this:
1. why match again, if the matching is already done
2. how is this safe against repeated patterns?

This could definitely use some comments.

Reply:
1.
lets say we have following string:

```
a = """hello{

      this.is.the.string

}"""
```

The first matching gives use:

```
match[1] = "{{

      this.is.the.string

}}"

```
```
match[2] ="                 #match 1 without the {{}}

      this.is.the.string

"
```


`match[3] = "this.is.the.string" #the actual variables`

With the second matching I look for the position of match[3] inside of match[2]. (knowing the variables should be on the same line)
this is done by:

- Splitting match[2] on linebreaks: ['','',"  `  ``  ``  `this.is.the.string",'']
- Iterating on each element of the array and with a second matching look for match[3] in the element.
- The index where it is found is the number of lines to add to startline of match[2]. and match.start() is the offset caused by the spaces that could be present before the variable.

2.As we only handle one {{}} at the time and it(match[2]) only contains spaces, linebreaks and the match, the match should never be found more then once.

Reply:
The reason we didn't is that it isn't necessarily correct: depending on what line endings the file uses you could get incorrect location reporting (`\n`, `\r`, `\r\n`). Since `splitlines` has all the know-how to handle this I thought it would be better to go with that. But I am convinced it can be made almost as readable with `splitlines` as with newline counts.

Reply:
@sanderr I think all the code in the parser is built on the assumption newline is '\n' 

```
def t_newline(t: lex.LexToken) -> None:  # noqa: N802
    r"\n+"
```

Reply:
@FloLey 
https://docs.python.org/3/library/re.html#re.Match.start

so it is not `match[2].start()` but `match.start(2)`

Reply:
Ho this is nice! thanks I'll try it out

Reply:
@sanderr @wouterdb I applied the changed removing the splitlines and using the idea of Wouter. It is a lot more readable.
(The tests are passing on the last build, just an issue with flake8 with should be fixed with my last commit)